{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "f4d1cabb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This notebook was prepared using:\n",
      "python version 3.10.9 | packaged by conda-forge | (main, Feb  2 2023, 20:26:08) [Clang 14.0.6 ]\n",
      "numpy version 1.24.2\n",
      "pandas version 1.5.3\n",
      "matplotlib version 3.7.1\n"
     ]
    }
   ],
   "source": [
    "print('This notebook was prepared using:')\n",
    "\n",
    "import os\n",
    "import sys\n",
    "print('python version %s' % sys.version)\n",
    "\n",
    "import numpy as np\n",
    "print('numpy version %s' % np.__version__)\n",
    "\n",
    "import pandas as pd\n",
    "print('pandas version %s' % pd.__version__)\n",
    "\n",
    "import math\n",
    "from math import isnan\n",
    "\n",
    "import matplotlib\n",
    "import matplotlib.cm as cm\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.gridspec as gridspec\n",
    "import matplotlib.image as mpimg\n",
    "print('matplotlib version %s' % matplotlib.__version__)\n",
    "\n",
    "import re\n",
    "import sys\n",
    "import argparse\n",
    "import scipy as sp\n",
    "try:\n",
    "    import itertools.izip as zip\n",
    "except ImportError:\n",
    "    import itertools\n",
    "import random\n",
    "\n",
    "from scipy import integrate\n",
    "import scipy.interpolate as sp_interpolate\n",
    "import statistics\n",
    "\n",
    "from dataclasses import dataclass\n",
    "import time as time_module\n",
    "\n",
    "# GitHub\n",
    "# HIV_DIR = 'data/HIV200'\n",
    "HIV_DIR = 'data/HIV'\n",
    "SIM_DIR = 'data/simulation'\n",
    "FIG_DIR = 'figures'\n",
    "MPL_DIR = 'src'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81068f41",
   "metadata": {},
   "source": [
    "1. all epitopes just had one or two escape mutations (700010077-5,700010058-5,700010040-5,706010164-5,704010042-5)\n",
    "2. all epitopes have more than two escape mutations (700010077-3,700010058-3,700010607-3,705010198-3,705010185-5,703010159-3) \n",
    "3. combination of the above (700010470-3 ,700010470-5 ,700010040-3,706010164-3, 705010162-3, 705010162-5, 704010042-3, 703010256-3, 703010256-5, 703010131-3, 703010131-5)\n",
    "4. no mutation related to escape (700010607-5,705010198-5,705010185-3,703010159-5)\n",
    "\n",
    "tag|sequence<br>length|escape<br>group|variants|max<br>time|time<br>step|gamma<br>s|raw<br>time|special<br>sites\n",
    ":----:|:----:|:---:|:-----:|:----|:----|:----:|:----:|:----:\n",
    "700010040-3|303|3|314|552|20|0.018|[0, 16, 45, 111, 181, 283, 412, 552]|[245, 246]\n",
    "700010040-5|146|0|146|552|5 |0.018|[0, 16, 45, 111, 181, 283, 412, 552]|[31, 32, 39, 104]\n",
    "700010058-3|90 |1|91 |85 |1 |0.118|[0, 8, 45, 85]|[]\n",
    "700010058-5|96 |0|96 |350|1 |0.029|[0, 8, 45, 85, 154, 239, 252, 350]|[13, 16, 17]\n",
    "700010077-3|203|6|221|159|5 |0.063|[0, 14, 32, 102, 159]|[]\n",
    "700010077-5|48 |0|49 |159|1 |0.063|[0, 14, 32, 159]|[1, 4, 5, 47]\n",
    "700010470-3|367|3|405|420|20|0.024|[0, 13, 41, 69, 174, 420]|[40, 334]\n",
    "700010470-5|193|3|198|454|5 |0.022|[0, 13, 41, 69, 174, 420, 454]|[8, 9, 124]\n",
    "700010607-3|239|1|252|21 |1 |0.476|[0, 9, 14, 21]|[]\n",
    "700010607-5|78 |0|78 |21 |1 |     |[0, 9, 14, 21]|no mutation related to epitope\t\n",
    "703010131-3|744|1|807|333|20|0.030|[0, 21, 28, 34, 63, 91, 175, 273, 333]|[0, 45, 46, 47, 76, 189, 560, 561]\n",
    "703010131-5|261|1|265|333|5 |0.030|[0, 21, 28, 34, 63, 91, 175, 273, 333]|[259]\n",
    "703010159-3|477|2|517|302|20|0.033|[0, 8, 12, 22, 29, 56, 85, 302]|[]\n",
    "703010159-5|216|0|224|302|5 |     |[0, 8, 12, 22, 29, 56, 85, 302]|no mutation related to epitope\t\n",
    "703010256-3|463|4|510|684|20|0.015|[0, 28, 63, 172, 426, 684]|[45, 46, 316, 317]\n",
    "703010256-5|402|1|410|684|20|0.015|[0, 28, 63, 172, 426, 684]|[42, 208]\n",
    "704010042-3|875|2|1074|676|20|0.015|[0, 21, 60, 172, 424, 676]|[95, 96, 765, 766]\n",
    "704010042-5|266|0|272|676|5|0.015|[0, 21, 60, 172, 424, 676]|[29, 30, 125]\n",
    "705010162-3|508|2|562|438|20|0.023|[0, 21, 77, 179, 438]|[33, 34, 327, 328]\n",
    "705010162-5|254|3|259|438|5 |0.023|[0, 21, 77, 179, 438]|[155]\n",
    "705010185-3|292|0|303|416|5 |     |[0, 25, 67, 180, 416]|no mutation related to epitope\t\n",
    "705010185-5|85 |1|89 |67 |1 |0.149|[0, 25, 67]|[]\n",
    "705010198-3|204|1|223|60 |1 |0.167|[0, 11, 60]|[]\n",
    "705010198-5|72 |0|73 |60 |1 |     |[0, 11, 60]|no mutation related to epitope\t\n",
    "706010164-3|485|3|532|434|20|0.023|[0, 14, 28, 70, 183, 434]|[463]\n",
    "706010164-5|204|0|205|434|5 |0.023|[0, 14, 28, 70, 183, 434]|[44, 50, 51, 170]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c912d699",
   "metadata": {},
   "source": [
    "Write sh file for HIV data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "e8965821",
   "metadata": {},
   "outputs": [],
   "source": [
    "infer_tags = ['700010040-3','700010040-5','700010058-3','700010058-5','700010077-3','700010077-5',\n",
    "              '700010470-3','700010470-5','700010607-3','703010131-3','703010131-5','703010159-3',\n",
    "              '703010256-3','703010256-5','704010042-5','705010162-3','705010162-5',\n",
    "              '705010185-5','705010198-3','706010164-3','706010164-5'] #'704010042-3'\n",
    "\n",
    "f = open(\"%s/HIV_run.sh\"%MPL_DIR,'w')\n",
    "f.write('#!/bin/bash\\n')\n",
    "\n",
    "# f.write('tags=(')\n",
    "# for tag in infer_tags:\n",
    "#     f.write('\"%s\" '%tag)\n",
    "# f.write(')\\n')\n",
    "# f.write('for tag in \"${tags[@]}\"\\n')\n",
    "# f.write('do\\n')\n",
    "# f.write('   python ../inference_HIV.py -tag %s --raw & \\n'%(tag))\n",
    "# f.write('done\\n')\n",
    "# f.write('wait\\n')\n",
    "# f.write('echo \"All scripts have been attempted.\"')\n",
    "# f.close()  \n",
    "\n",
    "for i in range(len(infer_tags)):\n",
    "    tag = infer_tags[i]\n",
    "    f.write('\\tpython ../inference_HIV.py -tag %s --raw'%(tag))\n",
    "    f.write(' || echo \"CH%s failed, continuing with next script...\" \\n'%tag[-5:])\n",
    "f.write('echo \"All scripts have been attempted.\"')\n",
    "f.close()  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e521a00d",
   "metadata": {},
   "source": [
    "Write sh file for HIV data with shorter time"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3a1fe919",
   "metadata": {},
   "source": [
    "\n",
    "tag|sequence<br>length|escape<br>group|variants|max<br>time|time<br>step|gamma<br>s|raw<br>time|special<br>sites\n",
    ":----:|:----:|:---:|:-----:|:----|:----|:----:|:----:|:----:\n",
    "700010470-3|282|3|301|174|5|0.057|[0, 13, 41, 69, 174]|[32]|\n",
    "700010470-5|139|1|141|174|5|0.057|[0, 13, 41, 69, 174]|[8, 17, 57, 59, 90]|\n",
    "703010159-3|312|2|322|85|1|0.118|[0, 8, 12, 22, 29, 56, 85]|[]|\n",
    "703010256-3|175|3|180|172|5|0.058|[0, 28, 63, 172]|[20, 105, 159, 160]|\n",
    "703010256-5|236|1|238|172|5|0.058|[0, 28, 63, 172]|[]|\n",
    "704010042-3|499|1|585|172|20|0.058|[0, 21, 60, 172]|[20, 21, 51, 52, 426]|\n",
    "704010042-5|80|0|80|172|1|0.058|[0, 21, 60, 172]|[9, 10]|\n",
    "705010162-3|394|2|404|179|20|0.056|[0, 21, 77, 179]|[18, 19, 224]|\n",
    "705010162-5|135|2|135|179|5|0.056|[0, 21, 77, 179]|[2, 3]|\n",
    "706010164-3|275|3|280|183|5|0.055|[0, 14, 28, 70, 183]|[269]|\n",
    "706010164-5|109|0|110|183|5|0.055|[0, 14, 28, 70, 183]|[27, 93]|\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06677284",
   "metadata": {},
   "outputs": [],
   "source": [
    "tags = ['700010470-3','700010470-5','703010159-3','703010256-3','703010256-5','704010042-3','704010042-5',\n",
    "        '705010162-3','705010162-5','706010164-3','706010164-5']\n",
    "\n",
    "f = open(\"%s/shorter.sh\"%MPL_DIR,'w')\n",
    "for tag in tags:\n",
    "    f.write('python ../inference_HIV_shorter.py -tag %s --raw\\n'%(tag))\n",
    "f.close()          \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6290405",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "661c816c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d879600",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d610af72",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb152915",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0514693d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d87c240f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e288de2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b266cc1e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "700010470-3|282|3|301|174|5|0.057|[0, 13, 41, 69, 174]|[32]|\n",
      "700010470-5|139|1|141|174|5|0.057|[0, 13, 41, 69, 174]|[8, 17, 57, 59, 90]|\n",
      "703010159-3|312|2|322|85|1|0.118|[0, 8, 12, 22, 29, 56, 85]|[]|\n",
      "703010256-3|175|3|180|172|5|0.058|[0, 28, 63, 172]|[20, 105, 159, 160]|\n",
      "703010256-5|236|1|238|172|5|0.058|[0, 28, 63, 172]|[]|\n",
      "704010042-3|499|1|585|172|20|0.058|[0, 21, 60, 172]|[20, 21, 51, 52, 426]|\n",
      "704010042-5|80|0|80|172|1|0.058|[0, 21, 60, 172]|[9, 10]|\n",
      "705010162-3|394|2|404|179|20|0.056|[0, 21, 77, 179]|[18, 19, 224]|\n",
      "705010162-5|135|2|135|179|5|0.056|[0, 21, 77, 179]|[2, 3]|\n",
      "706010164-3|275|3|280|183|5|0.055|[0, 14, 28, 70, 183]|[269]|\n",
      "706010164-5|109|0|110|183|5|0.055|[0, 14, 28, 70, 183]|[27, 93]|\n"
     ]
    }
   ],
   "source": [
    "NUC = ['-', 'A', 'C', 'G', 'T']\n",
    "\n",
    "tags = ['700010470-3','700010470-5','703010159-3','703010256-3','703010256-5','704010042-3','704010042-5',\n",
    "        '705010162-3','705010162-5','706010164-3','706010164-5']\n",
    "\n",
    "for i in range(len(tags)):\n",
    "    tag = tags[i]\n",
    "    result = AnalyzeData(tag)\n",
    "    print('%s|%d|%d|%d|%d|%d|%s|'%(tag,result.seq_length,len(result.escape_group),result.variants,\\\n",
    "                                   result.uniq_t[-1],result.time_step,result.gamma),end='')\n",
    "    print([int(i) for i in result.uniq_t],end='')\n",
    "    print('|',end='')\n",
    "    print([int(i) for i in result.special_sites],end='')\n",
    "    print('|')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e07d901b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from dataclasses import dataclass\n",
    "\n",
    "@dataclass\n",
    "class Result:\n",
    "    variants: 0\n",
    "    seq_length: 0\n",
    "    special_sites: []\n",
    "    uniq_t:[]\n",
    "    time_step:0\n",
    "    gamma:0\n",
    "    escape_group:[]\n",
    "    escape_TF:[]\n",
    "    trait_dis:[]\n",
    "    IntTime:[]\n",
    "        \n",
    "def AnalyzeData(tag,gamma1):\n",
    "    df_info = pd.read_csv('%s/analysis/%s-analyze.csv' %(HIV_DIR,tag), comment='#', memory_map=True)\n",
    "    seq     = np.loadtxt('%s/sequence/%s-poly-seq2state.dat'%(HIV_DIR,tag))\n",
    "\n",
    "    \"\"\"get raw time points\"\"\"\n",
    "    times = []\n",
    "    for i in range(len(seq)):\n",
    "        times.append(seq[i][0])\n",
    "    uniq_t = np.unique(times)\n",
    "\n",
    "    \"\"\"get variants number and sequence length\"\"\"\n",
    "    df_poly  = df_info[df_info['nucleotide']!=df_info['TF']]\n",
    "    variants = len(df_poly)\n",
    "    seq_length = df_info.iloc[-1].polymorphic_index + 1\n",
    "\n",
    "    \"\"\"get special sites and escape sites\"\"\"\n",
    "    # get all epitopes for one tag\n",
    "    df_rows = df_info[df_info['epitope'].notna()]\n",
    "    unique_epitopes = df_rows['epitope'].unique()\n",
    "\n",
    "    min_n = 2 # the least escape sites a trait group should have (more than min_n)\n",
    "    special_sites = [] # special site considered as time-varying site but not escape site\n",
    "    escape_group  = [] # escape group (each group should have more than 2 escape sites)\n",
    "    escape_TF     = [] # corresponding wild type nucleotide\n",
    "    for epi in unique_epitopes:\n",
    "        df_e = df_rows[(df_rows['epitope'] == epi) & (df_rows['escape'] == True)] # find all escape mutation for one epitope\n",
    "        unique_sites = df_e['polymorphic_index'].unique()\n",
    "\n",
    "        if len(unique_sites) <= min_n:\n",
    "            special_sites.append(unique_sites)\n",
    "        else:\n",
    "            escape_group.append(list(unique_sites))\n",
    "            tf_values = []\n",
    "            for site in unique_sites:\n",
    "                tf_value = df_e[df_e['polymorphic_index'] == site]['TF'].values\n",
    "                tf_values.append(NUC.index(tf_value[0]))\n",
    "            escape_TF.append(tf_values)\n",
    "\n",
    "    special_sites = [item for sublist in special_sites for item in sublist]\n",
    "\n",
    "    \"\"\"trait distance\"\"\"\n",
    "    trait_dis = []\n",
    "    if len(escape_group) > 0:\n",
    "        df_sequence = pd.read_csv('%s/processed/%s-index.csv' %(HIV_DIR,tag), comment='#', memory_map=True,usecols=['alignment','polymorphic'])\n",
    "        for i in range(len(escape_group)):\n",
    "            i_dis = []\n",
    "            for j in range(len(escape_group[i])-1):\n",
    "                index0 = df_sequence[df_sequence['polymorphic']==escape_group[i][j]].iloc[0].alignment\n",
    "                index1 = df_sequence[df_sequence['polymorphic']==escape_group[i][j+1]].iloc[0].alignment\n",
    "                i_dis.append(int(index1-index0))\n",
    "            trait_dis.append(i_dis)\n",
    "\n",
    "    \"\"\"find proper time step\"\"\"\n",
    "    if uniq_t[-1] < 100:\n",
    "        time_step = 1\n",
    "    else:\n",
    "        if seq_length > 300:\n",
    "            time_step = 20\n",
    "        elif seq_length < 100:\n",
    "            time_step = 1\n",
    "        else:\n",
    "            time_step = 5\n",
    "\n",
    "    \"\"\"find proper gamma value\"\"\"\n",
    "    # interpolation time according to the time step get above, all the inserted time points are integer\n",
    "    times = [0]\n",
    "    for t in range(1,len(uniq_t)):\n",
    "        tp_i   = round((uniq_t[t]-uniq_t[t-1])/time_step) # number of insertion points\n",
    "        if tp_i > 0:\n",
    "            ts_i   = round((uniq_t[t]-uniq_t[t-1])/tp_i) # modified time step\n",
    "            for i in range(tp_i):\n",
    "                time_i = uniq_t[t-1] + (i+1) * ts_i\n",
    "                time_s = uniq_t[t]\n",
    "                # if the inserted time is close to the bounadry time point, throw out the inserted point\n",
    "                if abs(time_i-time_s) <= ts_i/2:\n",
    "                    times.append(int(time_s))\n",
    "                else:\n",
    "                    times.append(int(time_i))\n",
    "        else:\n",
    "            times.append(uniq_t[t])\n",
    "    IntTime = list(times)\n",
    "\n",
    "    # regularize value gamma1, restricting the magnitude of the selection coefficients\n",
    "    gamma = round(gamma1/uniq_t[-1],3) # constant MPL gamma value / max time\n",
    "\n",
    "    return Result(variants, seq_length, special_sites, uniq_t, time_step, gamma, escape_group, escape_TF, trait_dis, IntTime)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "10765ee6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# loading data from dat file\n",
    "def getSequence(history,escape_TF,escape_group):\n",
    "    sVec      = []\n",
    "    nVec      = []\n",
    "    eVec      = []\n",
    "\n",
    "    temp_sVec   = []\n",
    "    temp_nVec   = []\n",
    "    temp_eVec   = []\n",
    "\n",
    "    times       = []\n",
    "    time        = 0\n",
    "    times.append(time)\n",
    "\n",
    "    ne          = len(escape_group)\n",
    "\n",
    "    for t in range(len(history)):\n",
    "        if history[t][0] != time:\n",
    "            time = history[t][0]\n",
    "            times.append(int(time))\n",
    "            sVec.append(temp_sVec)\n",
    "            nVec.append(temp_nVec)\n",
    "            eVec.append(temp_eVec)\n",
    "            temp_sVec   = []\n",
    "            temp_nVec   = []\n",
    "            temp_eVec   = []\n",
    "\n",
    "        temp_nVec.append(history[t][1])\n",
    "        temp_sVec.append(history[t][2:])\n",
    "\n",
    "        if ne > 0: # the patient contains escape group\n",
    "            temp_escape = np.zeros(ne, dtype=int)\n",
    "            for n in range(ne):\n",
    "                for nn in range(len(escape_group[n])):\n",
    "                    index = escape_group[n][nn] + 2\n",
    "                    if history[t][index] != escape_TF[n][nn]:\n",
    "                        temp_escape[n] = 1\n",
    "                        break\n",
    "            temp_eVec.append(temp_escape)\n",
    "\n",
    "        if t == len(history)-1:\n",
    "            sVec.append(temp_sVec)\n",
    "            nVec.append(temp_nVec)\n",
    "            eVec.append(temp_eVec)\n",
    "\n",
    "    return sVec,nVec,eVec\n",
    "\n",
    "# get muVec\n",
    "def getMutantS(sVec):\n",
    "    # use muVec matrix to record the index of time-varying sites\n",
    "    muVec = -np.ones((seq_length, q)) # default value is -1, positive number means the index\n",
    "    x_length  = 0\n",
    "\n",
    "    for i in range(seq_length):            \n",
    "        # find all possible alleles in site i\n",
    "        alleles     = [int(sVec[t][k][i]) for t in range(len(sVec)) for k in range(len(sVec[t]))]\n",
    "        allele_uniq = np.unique(alleles)\n",
    "        for allele in allele_uniq:\n",
    "            muVec[i][int(allele)] = x_length\n",
    "            x_length += 1\n",
    "\n",
    "    return x_length,muVec\n",
    "\n",
    "# calculate single and pair allele frequency (multiple case)\n",
    "def get_allele_frequency(sVec,nVec,eVec,muVec):\n",
    "\n",
    "    x  = np.zeros((len(nVec),x_length))           # single allele frequency\n",
    "    xx = np.zeros((len(nVec),x_length,x_length))  # pair allele frequency\n",
    "    for t in range(len(nVec)):\n",
    "        pop_size_t = np.sum([nVec[t]])\n",
    "        for k in range(len(nVec[t])):\n",
    "            # individual locus part\n",
    "            for i in range(seq_length):\n",
    "                qq = int(sVec[t][k][i])\n",
    "                aa = int(muVec[i][qq])\n",
    "                if aa != -1: # if aa = -1, it means the allele does not exist\n",
    "                    x[t,aa] += nVec[t][k]\n",
    "                    for j in range(int(i+1), seq_length):\n",
    "                        qq = int(sVec[t][k][j])\n",
    "                        bb = int(muVec[j][qq])\n",
    "                        if bb != -1:\n",
    "                            xx[t,aa,bb] += nVec[t][k]\n",
    "                            xx[t,bb,aa] += nVec[t][k]\n",
    "            # escape part\n",
    "            for n in range(ne):\n",
    "                aa = int(x_length-ne+n)\n",
    "                x[t,aa] += eVec[t][k][n] * nVec[t][k]\n",
    "                for m in range(int(n+1), ne):\n",
    "                    bb = int(x_length-ne+m)\n",
    "                    xx[t,aa,bb] += eVec[t][k][n] * eVec[t][k][m] * nVec[t][k]\n",
    "                    xx[t,bb,aa] += eVec[t][k][n] * eVec[t][k][m] * nVec[t][k]\n",
    "                for j in range(seq_length):\n",
    "                    qq = int(sVec[t][k][j])\n",
    "                    bb = int(muVec[j][qq])\n",
    "                    if bb != -1:\n",
    "                        xx[t,bb,aa] += eVec[t][k][n] * nVec[t][k]\n",
    "                        xx[t,aa,bb] += eVec[t][k][n] * nVec[t][k]\n",
    "        x[t,:]    = x[t,:]/pop_size_t\n",
    "        xx[t,:,:] = xx[t,:,:]/pop_size_t\n",
    "    return x,xx\n",
    "\n",
    "# calculate escape frequency\n",
    "def get_escape_fre_term(sVec,nVec):\n",
    "    ex  = np.zeros((len(nVec),ne,seq_length,q))\n",
    "    for t in range(len(nVec)):\n",
    "        pop_size_t = np.sum([nVec[t]])\n",
    "        for k in range(len(sVec[t])):\n",
    "            for n in range(ne):\n",
    "                n_mutations = 0\n",
    "                for nn in escape_group[n]:\n",
    "                    index = escape_group[n].index(nn)\n",
    "                    WT = escape_TF[n][index]\n",
    "                    if sVec[t][k][nn] != WT:\n",
    "                        n_mutations += 1\n",
    "                        site = nn\n",
    "                if n_mutations == 1:\n",
    "                    qq = int(sVec[t][k][site])\n",
    "                    ex[t,n,site,qq] += nVec[t][k]\n",
    "        ex[t,:,:,:] = ex[t,:,:,:] / pop_size_t\n",
    "    return ex\n",
    "\n",
    "# calculate frequencies for recombination part (binary case)  \n",
    "def get_p_k(sVec,nVec,seq_length,escape_group,escape_TF):\n",
    "    p_mut_k   = np.zeros((len(nVec),seq_length,3)) # 0: time, 1: all k point, 2: p_k, p_k-, p_k+\n",
    "    p_wt      = np.zeros((len(nVec),len(escape_group))) # 0: time, 1: escape group\n",
    "\n",
    "    for t in range(len(nVec)):\n",
    "        pop_size_t = np.sum([nVec[t]])\n",
    "\n",
    "        for n in range(len(escape_group)):\n",
    "            escape_group_n = escape_group[n]\n",
    "            sWT_n     = [int(i) for i in escape_TF[n]]\n",
    "\n",
    "            for k in range(len(sVec[t])): # different sequences at time t\n",
    "                sVec_n = [int(sVec[t][k][i]) for i in escape_group_n]\n",
    "\n",
    "                # no mutation within the trait group\n",
    "                if sWT_n == sVec_n:\n",
    "                    p_wt[t][n] += nVec[t][k]\n",
    "\n",
    "                for nn in range(len(escape_group_n)-1):\n",
    "                    k_bp = nn + 1\n",
    "\n",
    "                    # containing mutation before and after break point k,p_k\n",
    "                    if sWT_n[:k_bp] != sVec_n[:k_bp] and sWT_n[k_bp:] != sVec_n[k_bp:]:\n",
    "                        p_mut_k[t][escape_group_n[0]+nn][0] += nVec[t][k]\n",
    "\n",
    "                    # MT before break point k and WT after break point k,p_k-\n",
    "                    if sWT_n[:k_bp] != sVec_n[:k_bp] and sWT_n[k_bp:] == sVec_n[k_bp:]:\n",
    "                        p_mut_k[t][escape_group_n[0]+nn][1] += nVec[t][k]\n",
    "\n",
    "                    # WT before break point k and MT after break point k,p_k+\n",
    "                    if sWT_n[:k_bp] == sVec_n[:k_bp] and sWT_n[k_bp:] != sVec_n[k_bp:]:\n",
    "                        p_mut_k[t][escape_group_n[0]+nn][2] += nVec[t][k]\n",
    "\n",
    "        p_wt[t]    = p_wt[t] / pop_size_t\n",
    "        p_mut_k[t] = p_mut_k[t] / pop_size_t\n",
    "\n",
    "    return p_wt,p_mut_k\n",
    "\n",
    "# calculate mutation flux term \n",
    "def get_mutation_flux(x,ex,muVec):\n",
    "    flux = np.zeros((len(x),x_length))\n",
    "    for t in range(len(x)):\n",
    "        for i in range(seq_length):\n",
    "            for a in range(q):\n",
    "                aa = int(muVec[i][a])\n",
    "                if aa != -1:\n",
    "                    for b in range(q):\n",
    "                        bb = int(muVec[i][b])\n",
    "                        if b != a:\n",
    "                            if bb != -1:\n",
    "                                flux[t,aa] +=  muMatrix[b][a] * x[t,bb] - muMatrix[a][b] * x[t,aa]\n",
    "                            else:\n",
    "                                flux[t,aa] += -muMatrix[a][b] * x[t,aa]\n",
    "        for n in range(ne):\n",
    "            for nn in range(len(escape_group[n])):\n",
    "                for a in range(q):\n",
    "                    WT = escape_TF[n][nn]\n",
    "                    index = escape_group[n][nn]\n",
    "                    if a != WT:\n",
    "                        flux[t, x_length-ne+n] += muMatrix[WT][a] * (1 - x[t,x_length-ne+n]) - muMatrix[a][WT] * ex[t,n,index,a]\n",
    "    return flux\n",
    "\n",
    "# calculate recombination flux term\n",
    "def get_recombination_flux(x,p_wt,p_mut_k,trait_dis):\n",
    "    flux = np.zeros((len(x),x_length))\n",
    "    for n in range(ne):\n",
    "        for t in range(len(x)):\n",
    "            fluxIn  = 0\n",
    "            fluxOut = 0\n",
    "\n",
    "            for nn in range(len(escape_group[n])-1):\n",
    "                k_index = escape_group[n][0]+nn\n",
    "                fluxIn  += trait_dis[n][nn] * p_wt[t][n]*p_mut_k[t][k_index][0]\n",
    "                fluxOut += trait_dis[n][nn] * p_mut_k[t][k_index][1]*p_mut_k[t][k_index][2]\n",
    "\n",
    "            flux[t,x_length-ne+n] = r_rate * (fluxIn - fluxOut)\n",
    "\n",
    "    return flux\n",
    "\n",
    "# diffusion matrix C\n",
    "def diffusion_matrix_at_t(x,xx):\n",
    "    C = np.zeros([len(x),x_length,x_length])\n",
    "    for t in range(len(x)):\n",
    "        for i in range(x_length):\n",
    "            C[t,i,i] = x[t,i] - x[t,i] * x[t,i]\n",
    "            for j in range(int(i+1) ,x_length):\n",
    "                C[t,i,j] = xx[t,i,j] - x[t,i] * x[t,j]\n",
    "                C[t,j,i] = xx[t,i,j] - x[t,i] * x[t,j]\n",
    "    return C\n",
    "\n",
    "# calculate the difference between the frequency at time t and time t-1\n",
    "def cal_delta_x(single_freq,times):\n",
    "    delta_x = np.zeros((len(single_freq),x_length))   # difference between the frequency at time t and time t-1s\n",
    "#     calculate by np.gradient function\n",
    "#         for ii in range(x_length):\n",
    "#             delta_x[:,ii] = np.gradient(single_freq.T[ii],times)\n",
    "#     calculate manually\n",
    "    for t in range(len(single_freq)-1):\n",
    "        delta_x[t] = (single_freq[t+1] - single_freq[t])/(times[t+1]-times[t])\n",
    "    delta_x[-1] = delta_x[-2]\n",
    "    return delta_x\n",
    "\n",
    "# Interpolation function definition\n",
    "# Interpolates the input arrays so that they will have the same number of generations as the original population.\n",
    "interpolation = lambda a,b: sp_interpolate.interp1d(a,b,kind='linear')\n",
    "\n",
    "def interpolator_x(single_freq, double_freq, current_times, result_times):\n",
    "    single_freq_temp = np.zeros((len(result_times),x_length))\n",
    "    double_freq_temp = np.zeros((len(result_times),x_length,x_length))\n",
    "\n",
    "    for i in range(x_length):\n",
    "        single_freq_temp[:,i] = interpolation(current_times, single_freq[:,i])(result_times)\n",
    "        for j in range(x_length):\n",
    "            double_freq_temp[:,i,j] = interpolation(current_times, double_freq[:,i,j])(result_times)\n",
    "\n",
    "    return single_freq_temp, double_freq_temp\n",
    "\n",
    "def interpolator_ex(escape_freq, current_times, result_times):\n",
    "    if ne != 0:\n",
    "        escape_freq_temp  = np.zeros((len(result_times),ne,seq_length,q))\n",
    "    else: # if there is no escape group, return 0\n",
    "        escape_freq_temp  = []\n",
    "\n",
    "    for n in range(ne):\n",
    "        for i in range(seq_length):\n",
    "            for a in range(q):\n",
    "                escape_freq_temp[:,n,i,a] = interpolation(current_times, escape_freq[:,n,i,a])(result_times)\n",
    "\n",
    "    return escape_freq_temp\n",
    "\n",
    "def interpolator_p(p_wt, p_mut_k, current_times, result_times):\n",
    "    wt_temp    = np.zeros((len(result_times),ne))\n",
    "    mut_k_temp = np.zeros((len(result_times),seq_length,3))\n",
    "    # interpolation for wild type frequency\n",
    "    for n in range(ne):\n",
    "        wt_temp[:,n] = interpolation(current_times, p_wt[:,n])(result_times)\n",
    "\n",
    "    # interpolation for frequency related to recombination part\n",
    "    for i in range(seq_length):\n",
    "        for j in range(3):\n",
    "            mut_k_temp[:,i,j] = interpolation(current_times, p_mut_k[:,i,j])(result_times)\n",
    "\n",
    "    wt_temp    = wt_temp[:len(result_times)]\n",
    "    mut_k_temp = mut_k_temp[:len(result_times)]\n",
    "\n",
    "    return wt_temp, mut_k_temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "f507eb40",
   "metadata": {},
   "outputs": [],
   "source": [
    "tag = '700010058-3'\n",
    "\n",
    "data     = np.loadtxt(\"%s/sequence/%s-poly-seq2state.dat\" %(HIV_DIR,tag))\n",
    "\n",
    "# information for escape group\n",
    "result       = AnalyzeData(tag,10)\n",
    "escape_group = result.escape_group\n",
    "escape_TF    = result.escape_TF\n",
    "trait_dis    = result.trait_dis\n",
    "seq_length   = result.seq_length\n",
    "sample_times = result.uniq_t\n",
    "times        = result.IntTime\n",
    "ne           = len(escape_group)\n",
    "\n",
    "## regularization parameter\n",
    "p_sites      = result.special_sites\n",
    "time_step    = result.time_step\n",
    "gamma_1s     = result.gamma\n",
    "\n",
    "# obtain sequence data and frequencies\n",
    "sVec,nVec,eVec = getSequence(data,escape_TF,escape_group)\n",
    "x_length,muVec = getMutantS(sVec)\n",
    "x_length      += ne\n",
    "\n",
    "# get all frequencies, \n",
    "# x: single allele frequency, xx: pair allele frequency\n",
    "x,xx         = get_allele_frequency(sVec,nVec,eVec,muVec)\n",
    "\n",
    "# ex: escape frequency\n",
    "if ne != 0:\n",
    "    ex      = get_escape_fre_term(sVec,nVec)\n",
    "else:\n",
    "    ex      = 0\n",
    "\n",
    "# ex: escape frequency, p_wt,p_mut_k: frequency related to recombination part\n",
    "p_wt,p_mut_k = get_p_k(sVec,nVec,seq_length,escape_group,escape_TF)\n",
    "\n",
    "#record all input information before interpolation\n",
    "f = open('%s/rawdata/rawdata_%s.npz'%(HIV_DIR,tag), mode='w+b')\n",
    "escape_group = np.array(escape_group, dtype=object)\n",
    "escape_TF    = np.array(escape_TF , dtype=object)\n",
    "trait_dis    = np.array(trait_dis , dtype=object)\n",
    "np.savez_compressed(f, muVec=muVec, single_freq=x, double_freq=xx, escape_freq=ex, p_wt_freq=p_wt, p_mut_k_freq=p_mut_k,\\\n",
    "                    special_sites=p_sites, escape_group=escape_group, escape_TF=escape_TF,trait_dis=trait_dis,\\\n",
    "                    seq_length=seq_length, gamma = gamma_1s, time_step=time_step, sample_times=sample_times, times=times)\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "74f9a7ef",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Execution time for CH058-3 : 28.50273585319519 seconds\n"
     ]
    }
   ],
   "source": [
    "r_rate = 1.4e-5\n",
    "muMatrix = np.loadtxt(\"%s/Zanini-extended.dat\"%HIV_DIR)\n",
    "\n",
    "# load processed data from rawdata file\n",
    "\n",
    "rawdata  = np.load('%s/rawdata/rawdata_%s.npz'%(HIV_DIR,tag), allow_pickle=True)\n",
    "# information for individual sites\n",
    "x            = rawdata['single_freq']\n",
    "xx           = rawdata['double_freq']\n",
    "ex           = rawdata['escape_freq']\n",
    "muVec        = rawdata['muVec']\n",
    "sample_times = rawdata['sample_times']\n",
    "times        = rawdata['times']\n",
    "time_step    = rawdata['time_step']\n",
    "seq_length   = rawdata['seq_length']\n",
    "\n",
    "# information for escape group\n",
    "p_wt         = rawdata['p_wt_freq']\n",
    "p_mut_k      = rawdata['p_mut_k_freq']\n",
    "p_sites      = rawdata['special_sites']\n",
    "escape_group = rawdata['escape_group'].tolist()\n",
    "escape_TF    = rawdata['escape_TF'].tolist()\n",
    "\n",
    "ne           = len(escape_group)\n",
    "x_length     = len(x[0])\n",
    "\n",
    "# regularization value gamma_1 and gamma_2\n",
    "# individual site: gamma_1s, escape group: gamma_1p\n",
    "gamma_1s     = rawdata['gamma']\n",
    "gamma_1p     = gamma_1s/10\n",
    "gamma1 = np.ones(x_length)*gamma_1s\n",
    "for n in range(ne):\n",
    "    gamma1[x_length-ne+n] = gamma_1p\n",
    "\n",
    "# individual site: gamma_2c, escape group and special site: gamma_2tv\n",
    "gamma_2c  = 100000\n",
    "gamma_2tv = 500\n",
    "gamma2 = np.ones(x_length)*gamma_2c\n",
    "for n in range(ne):\n",
    "    gamma2[x_length-ne+n] = gamma_2tv\n",
    "for p_site in p_sites: # special site - time varying\n",
    "    for qq in range(len(NUC)):\n",
    "        index = int (muVec[p_site][qq]) \n",
    "        if index != -1:\n",
    "            gamma2[index] = gamma_2tv\n",
    "\n",
    "# after interpolation, calculate all the required data\n",
    "single_freq, double_freq = interpolator_x(x, xx, sample_times, times)\n",
    "escape_freq              = interpolator_ex(ex, sample_times, times)\n",
    "p_wt_freq, p_mut_k_freq  = interpolator_p(p_wt, p_mut_k,sample_times, times)\n",
    "\n",
    "# covariance matrix, flux term and delta_x\n",
    "covariance_n = diffusion_matrix_at_t(single_freq, double_freq)\n",
    "covariance   = np.swapaxes(covariance_n, 0, 2)\n",
    "flux_mu      = get_mutation_flux(single_freq,escape_freq,muVec)         # mutation part\n",
    "flux_rec     = get_recombination_flux(single_freq,p_wt_freq,p_mut_k_freq,trait_dis) # recombination part\n",
    "delta_x      = cal_delta_x(single_freq,times)\n",
    "\n",
    "# extend the time range\n",
    "TLeft   = int(round(times[-1]*0.5/10)*10)\n",
    "TRight  = int(round(times[-1]*0.5/10)*10)\n",
    "etleft  = np.linspace(-TLeft,-10,int(TLeft/10))\n",
    "etright = np.linspace(times[-1]+10,times[-1]+TRight,int(TRight/10))\n",
    "ExTimes = np.concatenate((etleft, times, etright))\n",
    "\n",
    "start_time = time_module.time()\n",
    "\n",
    "# solve the bounadry condition ODE to infer selections\n",
    "def fun(a,b):\n",
    "    \"\"\" Function defining the right-hand side of the system of ODE's\"\"\"\n",
    "    b_1                 = b[:x_length,:]   # the actual selection coefficients\n",
    "    b_2                 = b[x_length:,:]   # the derivatives of the selection coefficients, s'\n",
    "    result              = np.zeros((2*x_length,len(a))) # The RHS of the system of ODE's\n",
    "    result[:x_length]   = b_2       # sets the derivatives of the selection coefficients 'b_1', equal to s'\n",
    "    mat_prod            = np.sum(covariance[:,:,:len(a)] * b_1[:,len(etleft):len(etleft)+len(times)], 1)\n",
    "\n",
    "    for t in range(len(a)): # right hand side of second half of the ODE system\n",
    "        # within the time range\n",
    "        if len(etleft) <= t < len(etleft)+len(times):\n",
    "            tt = t - len(etleft)\n",
    "            for i in range(x_length):\n",
    "                result[x_length+i,t] = (mat_prod[i,tt] + gamma1[i] * b_1[i,t] + flux_mu[tt,i] + flux_rec[tt,i] - delta_x[tt,i]) / gamma2[i]\n",
    "\n",
    "        # outside the time range, no selection strength\n",
    "        else:\n",
    "            for i in range(x_length):\n",
    "                result[x_length+i,t] = gamma1[i] * b_1[i,t] / gamma2[i]\n",
    "\n",
    "    return result\n",
    "\n",
    "def fun_advanced(a,b):\n",
    "    \"\"\" The function that will be used if it is necessary for the BVP solver to add more nodes.\n",
    "    Note that the inference may be much slower if this has to be used.\"\"\"\n",
    "\n",
    "    b_1                 = b[:x_length,:]   # the actual selection coefficients\n",
    "    b_2                 = b[x_length:,:]   # the derivatives of the selection coefficients, s'\n",
    "    result              = np.zeros((2*x_length,len(a))) # The RHS of the system of ODE's\n",
    "    result[:x_length]   = b_2       # sets the derivatives of the selection coefficients 'b_1', equal to s'\n",
    "\n",
    "    # create new interpolated single and double site frequencies\n",
    "    single_freq_int, double_freq_int = interpolator_x(single_freq, double_freq, times, a)\n",
    "    escape_freq_int                  = interpolator_ex(escape_freq, times, a)\n",
    "    p_wt_int, p_mut_k_int            = interpolator_p(p_wt_freq,p_mut_k_freq,times,a,seq_length,ne)\n",
    "\n",
    "    # use the interpolations from above to get the values of delta_x and the covariance matrix at the nodes\n",
    "    flux_mu_int  = get_mutation_flux(single_freq_int, escape_freq_int, muVec) # mutation part\n",
    "    flux_rec_int = get_recombination_flux(single_freq_int,p_wt_int, p_mut_k_int, trait_dis)   # recombination part\n",
    "    delta_x_int  = cal_delta_x(single_freq_int,a)\n",
    "    covar_int    = diffusion_matrix_at_t(single_freq_int, double_freq_int)\n",
    "    covar_int    = np.swapaxes(covar_int,0,2)\n",
    "\n",
    "    # calculate the other half of the RHS of the ODE system\n",
    "    mat_prod_int  = np.sum(covar_int[:,:,:len(a)] * b_1[:,len(etleft):len(etleft)+len(times)], 1)\n",
    "\n",
    "    for t in range(len(a)): # right hand side of second half of the ODE system\n",
    "        # within the time range\n",
    "        if len(etleft) <= t < len(etleft)+len(times):\n",
    "            tt = t - len(etleft)\n",
    "            for i in range(x_length):\n",
    "                result[x_length+i,t] = (mat_prod_int[i,tt] + gamma1[i] * b_1[i,t] + flux_mu_int[tt,i] + flux_rec_int[tt,i] - delta_x_int[tt,i]) / gamma2[i]\n",
    "\n",
    "        # outside the time range, no selection strength\n",
    "        else:\n",
    "            for i in range(x_length):\n",
    "                result[x_length+i,t] = gamma1[i] * b_1[i,t] / gamma2[i]\n",
    "\n",
    "    return result\n",
    "\n",
    "# Boundary conditions\n",
    "# solution to the system of differential equation with the derivative of the selection coefficients zero at the endpoints\n",
    "def bc(b1,b2):\n",
    "    return np.ravel(np.array([b1[x_length:],b2[x_length:]])) # s' = 0 at the extended endpoints\n",
    "\n",
    "ss_extend = np.zeros((2*x_length,len(ExTimes)))\n",
    "\n",
    "try:\n",
    "    solution = sp.integrate.solve_bvp(fun, bc, ExTimes, ss_extend, max_nodes=10000, tol=1e-3)\n",
    "except ValueError:\n",
    "    print(\"BVP solver has to add new nodes\")\n",
    "    solution = sp.integrate.solve_bvp(fun_advanced, bc, ExTimes, ss_extend, max_nodes=10000, tol=1e-3)\n",
    "\n",
    "selection_coefficients = solution.sol(ExTimes)\n",
    "# removes the superfluous part of the array and only save the real time points\n",
    "desired_coefficients   = selection_coefficients[:x_length,len(etleft):len(etleft)+len(times)]\n",
    "\n",
    "end_time = time_module.time()\n",
    "print(f\"Execution time for CH{tag[6:]} : {end_time - start_time} seconds\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9617b6a2",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
